libfreerdp/core/proxy.c

```diff
@@ -253,3 +257,136 @@ BOOL http_proxy_connect(BIO* bufferedBio, const char* hostname, UINT16 port)
 	return TRUE;
 }
 
+/* SOCKS Proxy auth methods by rfc1928 */
+#define AUTH_M_NO_AUTH   0
+#define AUTH_M_GSSAPI    1
+#define AUTH_M_USR_PASS  2
+
+static BOOL socks_proxy_connect(BIO* bufferedBio, const char* hostname, UINT16 port)
+{
+	int status;
+	BYTE buf[280], hostnlen = 0xff & strlen(hostname);
+	/* CONN REQ replies in enum. order */
+	static const char *rplstat[] = {
+	  "succeeded",
+	  "general SOCKS server failure",
+	  "connection not allowed by ruleset",
+	  "Network unreachable",
+	  "Host unreachable",
+	  "Connection refused",
+	  "TTL expired",
+	  "Command not supported",
+	  "Address type not supported"
+	};
+
+	/* select auth. method */
+	memset(buf, '\0', sizeof(buf));
+	buf[0] = 5; /* SOCKS version */
+	buf[1] = 1; /* #of methods offered */
+	buf[2] = AUTH_M_NO_AUTH;
+	status = BIO_write(bufferedBio, buf, 3);
+	if (status != 3)
+	{
+		WLog_ERR(TAG, "SOCKS proxy: failed to write AUTH METHOD request");
+		return FALSE;
+	}
+
+	/* Read result SOCK reply */
+	memset(buf, '\0', sizeof(buf));
+	for(;;) {
+	  status = BIO_read(bufferedBio, buf, sizeof(buf));
```

Comment:
I guess you will run into troubles if the response packet is splitted in multiple small packets. So I guess you should loop until you have read a certain amount of bytes.
So in that sense, the memset on line 295 is not needed.

Reply:
reply on select auth req. is very short so the packet fragmentation can not happen here. It is also the first received packet so here the data are in sync.

Reply:
@yurashek Sadly that is up to the implementation, so better never assume anything beyond the indications returned.